### ISSUE TITLE
Provide a TAO-bypass for Navigation timing

### ISSUE THREAD

**Issue by yoavweiss** [w3c/resource-timing#220]
Navigation Timing is sharing many attribute definitions with Resource Timing, but doesn't use TAO, as it makes no sense in NT's context. We need to provide an internal flag that enables NT to reuse those definitions without requiring TAO. See https://github.com/w3c/navigation-timing/issues/117

**Comment by caribouW3** [comment]
Could you be more specific on what you mean by "it makes no sense in NT's context", since TAO plays a role for workerStart,
[w3c/navigation-timing#118](https://github.com/w3c/navigation-timing/issues/118)
The suggested flag would apply only to some definitions, right?

**Comment by yoavweiss** [comment]
The reason "it makes no sense" is that TAO checks if the destination origin is OK with exposing timing, where for NT, the destination origin is also the one with access to the data.
`workerStart` in the context of NT is always the origin's own SW (since it should be the [last one](https://github.com/w3c/navigation-timing/issues/100)), so it doesn't make sense of it to opt-in to itself.

Does that make sense?

**Comment by caribouW3** [comment]
Indeed, I had missed the clarification in issue [100](w3c/navigation-timing/issues/100). So that would add workerStart to the list in [117](w3c/navigation-timing/issues/117) I suppose.

**Comment by noamr** [comment]
Doesn't https://github.com/whatwg/html/pull/7105 fix this?

**Comment by annevk** [comment]
It does seem like in a redirect chain of A to B to C, A and B could opt-in to sharing their timing data. So C would get to know how these redirects impacted it. That's a slightly different model from TAO though so perhaps it needs its own design and header.

**Comment by noamr** [comment]
See https://github.com/w3c/navigation-timing/issues/160#issuecomment-1094828844

**Comment by noamr** [comment]
Suggesting one of two options:

- a new header that's similar to TAO: `Redirect-Timing-Allow-Origin`, which exposes this information and also in the future is an opt-in to get the navigation start time when cross-origin redirects are involved.
- A new header, perhaps `Timing-Allow-Context` or `Timing-Allow-Scope` with a list of values `resource | redirect | ...` - use the same TAO headers, but every time we want to add a new meaning to them such as redirects, we add a value to `Timing-Allow-Context`

**Comment by noamr** [comment]
OK, so a more detailed proposal based on the above and the WG discussion, to get this kicked off:
```
Resource-Timing-Policy: 1*<policy-directive>
<policy-directive>: <directive> <source>;
<directive>: resource | connection | exchange | redirect
<source>: <url-with-wildcards> / 'same-origin' / 'same-site' / 'none' / 'all' / 'nonce-{nonce}'
```

For example:
```
Resource-Timing-Policy: resource 'all'; connection 'same-site'; redirect '*.amazon.com'; exchange 'nonce-sha256-defasdgase123'
```

Each directive represents a group of attributes in resource timing:
```
redirect
  redirectStart
  redirectEnd
  // in the future: list of redirect URLs

connection
    nextHopProtocol;
    domainLookupStart;
    domainLookupEnd;
    connectStart;
    connectEnd;
    secureConnectionStart;

exchange
    fetchStart
    requestStart
    responseStart

resource
   encodedBodySize
   decodedBodySize
   transferSize
```

The following are always exposed, as they are either not TAO protected or anyway same-origin:
```
  initiatorType
  responseEnd
  workerStart // maybe this should move to navigation timing?
```

**Comment by yoavweiss** [comment]
Thanks @noamr for the great summary! I agree this is where we want to end up in the long term.

Random comments:
* At least some of the TAO-only attributes in your table should IMO also be enabled for CORS, as they are visible to `fetch()`.
* You left out `nextHopProtocol`, which is contentious, and may need to be in the same bucket as DNS timing.

Otherwise, I think we can split our path to get to that future to multiple chunks:
1) Expose the size attributes to CORS/CORP enabled content and estimate the loss of data of removing these attributes from TAO. Proceed with the latter carefully, perhaps while working with various 3P providers that currently TAO their content to also CORP-enable it.
2) Prototype aggregated perf metrics reporting with some of the primitives exposed as part of attribution reporting. Ideally, we won't need new primitives for it, but we might..
3) Once we have a concept of what aggregated reporting may look like, and it's something that's deployed for attribution reporting reasons, we can start exposing more info in those contexts and have RUM providers experiment with that.
4) Assuming experimentations are successful, we'd be able to successfully remove necessary attributes from TAO.

One risk with (3) would be implementation in some browsers but not others. I'm hoping that starting this conversation early can help us avoid that risk.

I'd love opinions of RUM providers on this path (@nicjansma, @cliffcrocker), as well as other vendors (@bdekoz, @sefeng211, @achristensen07), and web-app-sec folks (@mikewest, @camillelamy).

**Comment by camillelamy** [comment]
Thanks! I like the direction this proposal is going. In particular, removing `domainLookupStart` and `domainLookupEnd` from TAO is a good thing, as they relate to the state to the user's network, and it is not something that an origin should have access to.

**Comment by noamr** [comment]
> Thanks @noamr for the great summary! I agree this is where we want to end up in the long term.
> 
> Random comments:
> 
> * At least some of the TAO-only attributes in your table should IMO also be enabled for CORS, as they are visible to `fetch()`.

I think this is only true for `responseStart`. `requestStart` can expose connection times, it's not currently exposable via fetch AFAIK.

> * You left out `nextHopProtocol`, which is contentious, and may need to be in the same bucket as DNS timing.

Fixed

**Comment by yoavweiss** [comment]
> `requestStart` can expose connection times, it's not currently exposable via fetch AFAIK.

Agree on `requestStart`, although it doesn't necessarily equals the connectEnd time, as there could be delays between the two. 
OTOH, I believe `redirectEnd` and `fetchStart` are exposed to fetch (with "manual" redirect mode), so should be available to CORS responses.

**Comment by annevk** [comment]
It's not clear to me we want CORP to mean anything beyond "can be Spectre attacked". We had a long discussion about this elsewhere and I don't think we were able to come with a principled model around it.

**Comment by noamr** [comment]
> It's not clear to me we want CORP to mean anything beyond "can be Spectre attacked". We had a long discussion about this elsewhere and I don't think we were able to come with a principled model around it.

I'm actually of the same opinion, but wanted to keep that in mind in case that discussion resolves. The discussion was in this [Google Doc](https://docs.google.com/document/d/1JBUaX1xSOZRxBk5bRNZWgnzyJoCQC52TIRokACBSmGc/edit?resourcekey=0-cZ7da6v52enjwRSsp_tLyQ&disco=AAAAHiQKW4Q), took me a while to find it.

/cc @yoavweiss

**Comment by annevk** [comment]
It's also at #240 and I vaguely recall some other GitHub discussion, but I cannot find it easily.
